pkgs/by-name/fu/fuse-t/package.nix

```diff
@@ -0,0 +1,185 @@
+  # will need to be run with the environment variable `FUSE_NFSSRV_PATH`
+  # set to `<fuse-t.binaryDistribution>/bin/go-nfsv4`
+  binaryDistribution = stdenv.mkDerivation {
+    pname = "fuse-t-binary-dist";
+    inherit version;
+
+    src = fetchurl {
+      url = "https://github.com/macos-fuse-t/fuse-t/releases/download/${version}/fuse-t-macos-installer-${version}.pkg";
+      hash = "sha256-pYrQT05hGDcTPYXebTftvd3xGS8ozGTCa47FvWVZQ68=";
+    };
+
+    nativeBuildInputs = [
+      libarchive
+      p7zip
+      libtapi
+    ];
+    propagatedBuildInputs = [ DiskArbitration ];
+
+    unpackPhase = ''
+      7z x $src
+      bsdtar -xf Payload~
+    '';
+
+    buildPhase = ''
+      fuseDir="Library/Application Support/fuse-t"
+      sed -i "s|^prefix=.*|prefix=$out|" "$fuseDir/pkgconfig/fuse-t.pc"
+      sed -i "s|^\(Cflags: .*\)|\1 -D_FILE_OFFSET_BITS=64|" "$fuseDir/pkgconfig/fuse-t.pc"
+      ln -s libfuse-t-${version}.a "$fuseDir/lib/libfuse-t.a"
+      ln -s libfuse-t-${version}.dylib "$fuseDir/lib/libfuse-t.dylib"
+      rm "$fuseDir/uninstall.sh"
+    '';
+
+    installPhase = ''
+      runHook preInstall
+      mkdir -p $out/lib
+      mv "$fuseDir/bin" $out
+      mv "$fuseDir/include" $out
+      mv "$fuseDir/lib" $out
+      mv "$fuseDir/pkgconfig" $out/lib
+      mv Library $out/Library
+      runHook postInstall
+    '';
+
+    meta = {
+      description = "FUSE-T is a kext-less implementation of FUSE for macOS that uses NFS v4 local server instead of a kernel extension.";
+      homepage = "https://www.fuse-t.org/";
+      # free for non-commercial use, see: https://github.com/macos-fuse-t/fuse-t/blob/main/License.txt
+      license = [ lib.licenses.unfreeRedistributable ];
+      sourceProvenance = [ lib.sourceTypes.binaryNativeCode ];
+      platforms = lib.platforms.darwin;
+    };
+  };
+
+  stubs = stdenv.mkDerivation {
+    pname = "fuse-t-stubs";
+    inherit version;
+
+    src = fetchFromGitHub {
+      owner = "macos-fuse-t";
+      repo = "libfuse";
+      rev = "da72c8b60eb929d86c4594fc1792ae49865cd659";
+      hash = "sha256-++edBetY+Xfvp5X4i+mL+FJgJXeV43toDRRKhLHlMhw=";
+    };
+
+    patches = [
+      ./no-universal-binaries.patch
+      ./use-correct-path.patch
+    ];
+
+    postPatch = ''
+      substituteAllInPlace lib/fuse_darwin_private.h
+    '';
+
+    nativeBuildInputs = [
+      cmake
+      libtapi
+    ];
+
+    # TODO: remove `.out` after https://github.com/NixOS/nixpkgs/pull/346043 is merged to `master`
+    buildInputs = [
+      libiconv.out
+      DiskArbitration
+    ];
+
+    propagatedBuildInputs = [ DiskArbitration ];
+
+    # Removes a lot of warning spam
+    # `warning: 'pthread_setugid_np' is deprecated: Use of per-thread security contexts is error-prone and discouraged.`
+    env.NIX_CFLAGS_COMPILE = "-Wno-deprecated-declarations";
+
+    postBuild = ''
+      mkdir -p lib/pkgconfig
+      tapi stubify --filetype=tbd-v2 lib/libfuse-t.dylib -o lib/libfuse-t.tbd
```

Comment:
What's the purpose of creating a TBD stub from a source-built package? I'm confused because I recognize TBD stubs as substitutes for actual dylib files, yet the built dylib is included in this package.

Reply:
I just copied it from the macFUSE derivation, I didn't actually know what they were for

Reply:
I see. So here's the background. macFUSE consists of a kernel extension, preference pane, and the userspace libfuse library. This created a few challenges.

1. Everything except libfuse is unfree which disqualifies it from being the on binary cache
2. Preference panes and kernel extensions can't be installed with Nix
3. libfuse works closely with the kernel extension, so version mismatches could break things

So I took the approach of making users install macFUSE outside of Nix and have packages link to the out-of-store libfuse. To achieve this, the macfuse-stub package creates a stub of libfuse using the actual macFUSE release so that compilers can link to it without it being actually present in /usr/local/lib.

With FUSE-T, problem No. 1 remains. And if drop the NFS server but build libfuse from source, we'll have problem No. 3 as well. So I think the FUSE-T package should take the same approach to avoid these problems.

Reply:
In the case of FUSE-T, the location for the NFS server is compiled in a header https://github.com/NixOS/nixpkgs/pull/348804/files#diff-542ad4d614512f9a215eaba4a723df42cc4a1fdc6f1d339f73a35818a50b043e but can be overridden with the environment variable `FUSE_NFSSRV_PATH`

So we could change it from directly pointing to the unfree NFS server to `/run/current-system/sw/bin` and then nix-darwin users can install the unfree NFS server using `environment.systemPackages`

In addition we could create wrappers (living under different attributes, so the downstream packages still get cached) that set `FUSE_NFSSRV_PATH` so that users can still use the packages from Nixpkgs without needing to manually install the unfree NFS server

Reply:
I don’t suppose there’s any chance of using the same stubs for both macFUSE and FUSE-T?

Reply:
I don't think so. FUSE-T changed the filename to `libfuse-t.dylib` for one thing.

Reply:
> So we could change it from directly pointing to the unfree NFS server to /run/current-system/sw/bin and then nix-darwin users can install the unfree NFS server using environment.systemPackages

For maximum compatibility, I guess we can use the header as is and introduce a nix-darwin module that makes `FUSE_NFSSRV_PATH` point to the Nix store. This would allow both Nix and non-Nix applications to use FUSE-T.

It should be noted though, that some applications use libraries that bypass libfuse and talk to the NFS server directly. Those aren't guaranteed to respect `FUSE_NFSSRV_PATH`. This is likely not an exhaustive list, but I found one case where the path is hard coded.

https://grep.app/search?q=FUSE_NFSSRV_PATH

> In addition we could create wrappers (living under different attributes, so the downstream packages still get cached) that set FUSE_NFSSRV_PATH so that users can still use the packages from Nixpkgs without needing to manually install the unfree NFS server

You mean like `steam-run`? That sounds nice, though I think we'd need to check that two different instances of the NFS server can run at the same time. Nix being Nix, users may have multiple FUSE-T NFS servers in their Nix store.

From what I've seen in FUSE-T libfuse's code, it does appear to be okay. libfuse creates a socket pair using `socketpair()`, forks and launches the NFS server, and uses the created socket to communicate with the newly launched server. It's probably a good idea to check though, because we don't know exactly what happens in the NFS server.